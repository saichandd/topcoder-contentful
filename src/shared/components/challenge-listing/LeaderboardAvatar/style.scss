@import '~styles/mixins';
$leader-space-10: $base-unit * 2;
$leader-space-15: $base-unit * 3;
$leader-space-20: $base-unit * 4;
$leader-space-25: $base-unit * 5;
$leader-space-30: $base-unit * 6;
$leader-space-40: $base-unit * 8;
$leader-space-50: $base-unit * 10;
$leader-radius-50: $corner-radius * 25;
$leader-radius-4: $corner-radius * 2;

.leaderboard-avatar {
  position: relative;
  text-align: center;
  font-size: 12px;
  color: $tc-white;
  line-height: $leader-space-30;
  display: inline-block;
  width: $leader-space-30;
  height: $leader-space-30;
  border-radius: $leader-radius-50;
  vertical-align: text-bottom;

  img {
    width: 100%;
    height: 100%;
    border-radius: $leader-radius-50;
    vertical-align: top;
  }

  &.light-gray {
    background: $tc-gray-10;
    color: $tc-black;

    &:hover {
      background: $tc-gray-20;
    }
  }

  .placement {
    position: absolute;
    display: block;
    width: $leader-space-15 + 2;
    height: $leader-space-15 + 2;
    right: -($leader-space-10 - 3);
    bottom: -($leader-space-10 - 3);
    text-align: center;
    line-height: $leader-space-10 + 3;
    border-radius: $leader-radius-50;
    border: 2px solid $tc-white;

    &.placement-1 {
      background: $tc-gold;
      color: $tc-gold-130;
    }

    &.placement-2 {
      background: $tc-silver;
      color: $tc-silver-130;
    }

    &.placement-3 {
      background: $tc-bronze;
      color: $tc-bronze-130;
    }

    &.placement-4 {
      background: $tc-dark-blue-30;
      color: $tc-dark-blue-100;
    }
  }

  .hidden {
    display: none;
  }
}
