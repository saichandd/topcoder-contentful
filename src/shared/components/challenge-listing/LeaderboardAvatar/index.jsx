// TODO: All uses of this component MUST be replaced by Avatar component!

import { Avatar } from 'topcoder-react-ui-kit';
import React, { Component } from 'react';
import PT from 'prop-types';
import { config, Link } from 'topcoder-react-utils';
import './style.scss';

/* TODO: Should be functional component! */
class LeaderboardAvatar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      member: props.member,
    };
  }

  render() {
    const {
      onClick, openNewTab, plusOne, url,
    } = this.props;
    const { member } = this.state;
    const targetURL = url || `${config.URL.BASE}/members/${member.handle}`;
    let { photoURL } = member;
    if (photoURL) {
      /* Note: 50px is larger than we really need here (the avatar size is
       * 30px), but it matches the target avatar size in the tooltip, shown
       * by the challenge listing when an avatar is hovered. Having the same
       * size will ensure that the avatar in tooltip always appears immediately.
       */
      photoURL = `${config.CDN.PUBLIC}/avatar/${
        encodeURIComponent(photoURL)}?size=50`;
    }
    return (
      <Link
        enforceA
        onClick={onClick}
        to={targetURL}
        styleName={`leaderboard-avatar ${member.position || member.isSmr ? '' : 'light-gray'}`}
        target={openNewTab ? '_blank' : undefined}
      >
        { plusOne ? member.handle : <Avatar url={photoURL} />}
        <span styleName={member.position ? `placement placement-${member.position}` : 'hidden'}>
          {member.position}
        </span>
      </Link>
    );
  }
}

LeaderboardAvatar.defaultProps = {
  member: {},
  onClick: null,
  openNewTab: false,
  plusOne: false,
  url: '',
};

LeaderboardAvatar.propTypes = {
  member: PT.shape({}),
  onClick: PT.func,
  openNewTab: PT.bool,
  plusOne: PT.bool,
  url: PT.string,
};

export default LeaderboardAvatar;
