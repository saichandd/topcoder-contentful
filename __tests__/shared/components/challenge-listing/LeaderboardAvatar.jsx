import React from 'react';
import Renderer from 'react-test-renderer/shallow';
import TU from 'react-dom/test-utils';
import LeaderboardAvatar from 'components/challenge-listing/LeaderboardAvatar';
import { StaticRouter } from 'react-router-dom';

const mockData = {
  member: {
    handle: 'handle',
  },
  domain: 'domain',
  url: 'url',
};

const mockData2 = {
  member: {
    handle: 'handle',
    isSmr: true,
    photoURL: 'photoURL',
  },
  domain: 'domain',
};

const mockData3 = {
  member: {
    handle: 'handle',
    isSmr: true,
    photoURL: 'photoURL',
    position: '1',
  },
  domain: 'domain',
};

test('Matches shallow shapshot', () => {
  const renderer = new Renderer();

  renderer.render((
    <StaticRouter context={{}}>
      <LeaderboardAvatar {...mockData} />
    </StaticRouter>
  ));
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});

test.skip('Render properly', () => {
  let instance = TU.renderIntoDocument((
    <StaticRouter context={{}}>
      <LeaderboardAvatar {...mockData2} />
    </StaticRouter>
  ));
  /* TODO: This is a wrong way to make such checks: due to updates in the code
   * I had to wrap LeaderboardAvatar inside StaticRouter, and this check belows
   * breaks, as it reliese on the specific structure, rather than making a
   * proper search through the rendered component tree. */
  expect(instance.state.member).toEqual(mockData2.member);

  instance = TU.renderIntoDocument((
    <StaticRouter context={{}}>
      <LeaderboardAvatar {...mockData3} />
    </StaticRouter>
  ));
  expect(instance.state.member).toEqual(mockData3.member);
});
