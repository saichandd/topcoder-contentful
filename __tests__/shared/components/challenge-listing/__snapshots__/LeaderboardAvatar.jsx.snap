// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shallow shapshot 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <LeaderboardAvatar
    domain="domain"
    member={
      Object {
        "handle": "handle",
      }
    }
    onClick={null}
    openNewTab={false}
    plusOne={false}
    url="url"
  />
</Router>
`;
